hazelcast/src/hazelcast/client/metrics.cpp

```diff
@@ -0,0 +1,317 @@
+#include <string>
+#include <vector>
+
+#include <boost/endian/conversion.hpp>
+
+#include <zlib.h>
+
+#include <hazelcast/client/impl/metrics/metric_descriptor.h>
+#include <hazelcast/client/impl/metrics/metrics_compressor.h>
+#include <hazelcast/client/impl/metrics/metrics_dictionary.h>
+
+namespace hazelcast {
+namespace client {
+namespace impl {
+namespace metrics {
+
+namespace {
+
+constexpr int MAX_WORD_LENGTH = 255;
+constexpr int NULL_DICTIONARY_ID = -1;
+
+constexpr int MASK_PREFIX = 1 << 0;
+constexpr int MASK_METRIC = 1 << 1;
+constexpr int MASK_DISCRIMINATOR = 1 << 2;
+constexpr int MASK_DISCRIMINATOR_VALUE = 1 << 3;
+constexpr int MASK_UNIT = 1 << 4;
+constexpr int MASK_EXCLUDED_TARGETS = 1 << 5;
+constexpr int MASK_TAG_COUNT = 1 << 6;
+
+constexpr int SIZE_VERSION = 2;
+constexpr int SIZE_DICTIONARY_BLOB = 4;
+constexpr int SIZE_COUNT_METRICS = 4;
+
+constexpr int VALUE_TYPE_LONG = 0;
+constexpr int VALUE_TYPE_DOUBLE = 1;
+
+constexpr int BITS_IN_BYTE = 8;
+constexpr int BYTE_MASK = 0xff;
+constexpr int BINARY_FORMAT_VERSION = 1;
+
+
+std::size_t find_common_prefix_length(const std::string &s1, const std::string &s2) {
+    std::size_t len = 0;
+    while (len < s1.size() && len < s2.size() && s1[len] == s2[len]) {
+        len++;
+    }
+    return len;
+}
+
+std::vector<byte> zlib_compress(const std::vector<byte> &input) {
+    const auto max_output_len = compressBound(input.size());
+
+    auto output_len = max_output_len;
+    std::vector<byte> output(output_len);
+
+    int rc = compress2(output.data(), &output_len, input.data(), input.size(), Z_BEST_SPEED);
+
+    if (rc == Z_MEM_ERROR) {
+        throw std::bad_alloc{};
```

Comment:
why not use one of our library exceptions? we usually say that we will throw exception derived from `iexception`.

Reply:
bad_alloc is the exception you would get when you don't have enough memory. Hazelcast c++ client functions may as well throw this exception when memory is full.